### PR TITLE
Refactor permission policy RBAC

### PR DESCRIPTION
**Pull request by bernardo-bastos**
Refactors the permission policy to follow the following format: `<ROLE>, <ACTION>, <ENVIRONMENT_GROUP>:<ENVIRONMENT>, <APPLICATION>, allow`

The allowed actions are: `CreateLock`, `DeleteLock`, `CreateRelease`, `DeployRelease`, `CreateUndeploy`, `DeployUndeploy`, `CreateEnvironment`, `CreateEnvironmentApplication` and `DeployReleaseTrain`.

The following actions `CreateUndeploy`, `DeployUndeploy` and `CreateEnvironmentApplication` are environment independent meaning that the environment specified on the permission needs to follow the following format `<ENVIRONMENT_GROUP>:*`, otherwise an error will be thrown.

### CODE REVIEW

File: charts/kuberpult/values.yaml.tpl [hunk 1]
```diff
@@ -155,8 +155,13 @@ auth:
 dex:
   enabled: false
   # Defines the rbac policy when using Dex.
-  # The permissions are added using the following format (p, <ROLE>, <APPLICATION>, <ACTION>, <ENV>, <PERMISSION>), example:
-  # p, Developer, deployment, *, dev:development-d2, allow
+  # The permissions are added using the following format (<ROLE>, <ACTION>, <ENVIRONMENT_GROUP>:<ENVIRONMENT>, <APPLICATION>, allow). 
+  # 
+  # Available actions are: CreateLock, DeleteLock, CreateRelease, DeployRelease, CreateUndeploy, DeployUndeploy, CreateEnvironment, CreateEnvironmentApplication and DeployReleaseTrain.
+  # The following actions CreateUndeploy, DeployUndeploy and CreateEnvironmentApplication are environment independent meaning that the environment specified on the permission
```
Comment (sven-urbanski-freiheit-com) [inline charts/kuberpult/values.yaml.tpl:161]:
```suggestion
  # The actions CreateUndeploy, DeployUndeploy and CreateEnvironmentApplication are environment independent meaning that the environment specified on the permission 
```

File: charts/kuberpult/values.yaml.tpl [hunk 2]
```diff
@@ -155,8 +155,13 @@ auth:
...
   # Defines the rbac policy when using Dex.
-  # The permissions are added using the following format (p, <ROLE>, <APPLICATION>, <ACTION>, <ENV>, <PERMISSION>), example:
-  # p, Developer, deployment, *, dev:development-d2, allow
+  # The permissions are added using the following format (<ROLE>, <ACTION>, <ENVIRONMENT_GROUP>:<ENVIRONMENT>, <APPLICATION>, allow). 
+  # 
+  # Available actions are: CreateLock, DeleteLock, CreateRelease, DeployRelease, CreateUndeploy, DeployUndeploy, CreateEnvironment, CreateEnvironmentApplication and DeployReleaseTrain.
+  # The following actions CreateUndeploy, DeployUndeploy and CreateEnvironmentApplication are environment independent meaning that the environment specified on the permission 
+  # needs to follow the following format <ENVIRONMENT_GROUP>:*, otherwise an error will be thrown.
+  #
+  # Example permission: Developer, CreateLock, development:development, *, allow
```
Comment (sven-urbanski-freiheit-com) [inline charts/kuberpult/values.yaml.tpl:164]:
```suggestion
  # Example permission: Developer, CreateLock, development:development, *, allow
  # If no goup is configured for an environment, the environmet group name is the same as the environment name, here "development"
```

File: pkg/auth/rbac.go [hunk 3]
```diff
@@ -33,41 +33,33 @@ import (
 type RBACConfig struct {
 	// Indicates if Dex is enabled.
 	DexEnabled bool
-	// The RBAC policy. Key is for example "p,Developer,EnvironmentLock,Create,production,allow"
-	Policy map[string]*Permission
+	Policy     map[string]*Permission
```
Comment (sven-urbanski-freiheit-com) [inline pkg/auth/rbac.go:36]: why delete the comment? Either keep the comment or explain in other words what the key of this map is.

File: pkg/auth/rbac.go [hunk 4]
```diff
@@ -33,41 +33,33 @@ import (
...
 }
 
 // Inits the RBAC Config struct
 func initPolicyConfig() policyConfig {
 	return policyConfig{
-		allowedApps:    []string{"EnvironmentLock", "EnvironmentApplicationLock", "Deploy", "Undeploy", "EnvironmentFromApplication"},
-		allowedActions: []string{"Create", "Delete"},
+		// List of allowed actions on the RBAC policy.
+		allowedActions: []string{
+			"CreateLock",
```
Comment (sven-urbanski-freiheit-com) [inline pkg/auth/rbac.go:44]: These should be constants, e.g.
const (
	PermissionReleaseTrain = "ReleaseTrain"
)

File: pkg/auth/rbac.go [hunk 5]
```diff
@@ -76,30 +68,49 @@ func (c *policyConfig) validateAction(action string) error {
...
 	}
 	// Validate <ENVIRONMENT_GROUP:ENVIRONMENT>
-	if len(e) == 2 {
-		if !valid.EnvironmentName(e[0]) {
-			return fmt.Errorf("invalid environment group %s", envs)
-		}
-		if !valid.EnvironmentName(e[1]) {
-			return fmt.Errorf("invalid environment %s", envs)
-		}
+	if !valid.EnvironmentName(e[0]) {
```
Comment (sven-urbanski-freiheit-com) [inline pkg/auth/rbac.go:77]:
```suggestion
       envName := e[0]
       groupName :=e[1]
	if !valid.EnvironmentName(e[0]) {
```

File: pkg/auth/rbac.go [hunk 6]
```diff
@@ -112,29 +123,29 @@ func ValidateRbacPermission(line string) (p *Permission, err error) {
...
 	// Verifies if all fields are specified
 	c := strings.Split(line, ",")
-	if len(c) != 6 {
-		return nil, fmt.Errorf("6 fields are expected but only %d were specified", len(c))
+	if len(c) != 5 {
+		return nil, fmt.Errorf("5 fields are expected but only %d were specified", len(c))
 	}
-	// Validates the permission app
-	err = cfg.validateApp(c[2])
+	// Validates the permission action
```
Comment (sven-urbanski-freiheit-com) [inline pkg/auth/rbac.go:129]:
```suggestion
	// Validates the permission action
	action := c[1]
	// etc for the other fields, for readibility
```

File: charts/kuberpult/values.yaml.tpl (hunk 2 above)
Comment (bernardo-bastos) [reply charts/kuberpult/values.yaml.tpl:164]: Updated the comment.

File: pkg/auth/rbac.go (hunk 6 above)
Comment (bernardo-bastos) [reply pkg/auth/rbac.go:148]: Same here :)

File: pkg/auth/rbac.go (hunk 5 above)
Comment (bernardo-bastos) [reply pkg/auth/rbac.go]: Added local variables to make the code more readable.

File: pkg/auth/rbac.go [hunk 7]
```diff
@@ -76,30 +81,53 @@ func (c *policyConfig) validateAction(action string) error {
...
+	if !valid.EnvironmentName(groupName) {
+		return fmt.Errorf("invalid environment group %s", envs)
 	}
-	// Validate <ENVIRONMENT>
-	if len(e) == 1 {
-		if !valid.EnvironmentName(e[0]) {
-			return fmt.Errorf("invalid environment %s", envs)
+	// Actions that are environment independent need to follow the format <ENVIRONMENT_GROUP:*>.
+	if isEnvironmentIndependent(action) {
+		if e[1] == "*" {
```
Comment (sven-urbanski-freiheit-com) [inline pkg/auth/rbac.go:98]:
```suggestion
		if envName == "*" {
```

**Review by sven-urbanski-freiheit-com — APPROVED** [review]

**Pull request merged**